### PR TITLE
Merge main into feature/proto-upgrade

### PR DESCRIPTION
**Pull request by mzabaluev**
Contributes to #1330

Update the feature branch to the recent changes in main.

Of particular interest are the protobuf changes finalized in the
0.38.0 release, and the addition of the ADR-101 gRPC services.

### CODE REVIEW

**Comment by mzabaluev** [comment]
Commits starting from 30ddb516845c5ed0943963d20bec6a595818d60c are reviewable.

File: api/cometbft/consensus/v1/message.go [hunk 1]
```diff
@@ -38,12 +38,6 @@ func (m *HasVote) Wrap() proto.Message {
 	return cm
 }
 
-func (m *HasProposalBlockPart) Wrap() proto.Message {
```
Comment (sergio-mena) [inline api/cometbft/consensus/v1/message.go]: Not sure I understand this hunk

File: api/cometbft/consensus/v1/message.go [hunk 2]
```diff
@@ -105,9 +99,6 @@ func (m *Message) Unwrap() (proto.Message, error) {
 	case *Message_HasVote:
 		return m.GetHasVote(), nil
 
-	case *Message_HasProposalBlockPart:
```
Comment (sergio-mena) [inline api/cometbft/consensus/v1/message.go]: Same here

File: api/cometbft/consensus/v1/message.go (hunk 1 above)
Comment (sergio-mena) [reply api/cometbft/consensus/v1/message.go]: Got it. It was moved to `v2`

File: consensus/msgs.go [hunk 3]
```diff
@@ -7,81 +7,82 @@ import (
...
 	"github.com/cometbft/cometbft/types"
 )
 
-// Takes a consensus message type and returns the proto defined consensus message,
-// wrapped in the discriminating Message container.
-func MsgToWrappedProto(msg Message) (cmtcons.Message, error) {
-	pb := cmtcons.Message{}
+// MsgToProto takes a consensus message type and returns the proto defined consensus message.
+//
+// TODO: This needs to be removed, but WALToProto depends on this.
```
Comment (sergio-mena) [inline consensus/msgs.go:20]: Can we open an issue on this?

**Review by sergio-mena — APPROVED** [review]
Reviewed the merge commit as a 3-way diff.
Reviewed the rest of commits on this PR

It all looks good to me, thanks for this! 🙏

As this will land as a "merge", please collapse 30ddb51 and following in one commit, then (optionally) "amend" the merge commit with the resulting collapsed commit.

File: consensus/msgs.go (hunk 3 above)
Comment (mzabaluev) [reply consensus/msgs.go:20]: Sure, however this comment is present in main, I just merged it here.

**Comment by mzabaluev** [comment]
Some dynamically-typed code ended up with a wrong message type again, will fix.

File: consensus/msgs.go [hunk 4]
```diff
@@ -115,10 +115,10 @@ func MsgToProto(msg Message) (proto.Message, error) {
 			vsb.Votes = *bits
 		}
 
-		pb = vsb
+		pb.Sum = &cmtcons.Message_VoteSetBits{VoteSetBits: vsb}
 
 	default:
-		return nil, ErrConsensusMessageNotRecognized{msg}
+		return pb, ErrConsensusMessageNotRecognized{msg}
```
Comment (sergio-mena) [inline consensus/msgs.go:121]: Why this change? (from `nil` to `pb`)
If `pb` is `nil` in this code path, then `nil` is more readable

Comment (mzabaluev) [reply consensus/msgs.go:121]: The helper function got changed to return a struct value rather than a type-erased `proto.Message` interface pointer, eliminating a redundant roundtrip of a `Wrap` call and a dynamic downcast.

**Comment by mzabaluev** [comment]
The remaining failure looks like some jitter in pruning service tests.

**Review by sergio-mena — APPROVED** [review]
Re-approving

File: mempool/clist_mempool.go [hunk 5]
```diff
@@ -310,6 +315,9 @@ func (mem *CListMempool) globalCb(req *abci.Request, res *abci.Response) {
 			}
 			mem.metrics.RecheckTimes.Add(1)
 			mem.resCbRecheck(req, res)
+
+		default:
+			panic("unexpected value of RequestCheckTx.type")
```
[the remark this answers is not in the document]
Comment (mzabaluev) [reply mempool/clist_mempool.go]: This should never happen because the `Type` field is set elsewhere in the same module before the struct value is passed to the callback. Added the formatting just to be more informative.
